// echo/lib/backplane/BackplanePoller.js
var events = require('events'),
        http = require('http');
var utils = require('./../utils.js');

var inspect = require('sys').inspect;

var BackplanePoller = events.EventEmitter.prototype.spawn();

Object.defineProperties(BackplanePoller,{
    ctor: {
        value: function(args){
            if(args.ssl) this.client = http.createClient(443,args.host,true);
            else this.client = http.createClient(80,this.host,false);
            setInterval(this.bind(this.getBusMessages),this.interval);
        },
        writable: false,
        enumerable: true,
        configurable: true
    },
    client: {
        writable: false,
        enumerable: true,
        configurable: false
    },
    lastMessageId: {
        writable: false,
        enumerable: true,
        configurable: false
    },
    getBusMessages: {
        value: function(){
            var path =  '/v1/bus/' + this.busName;
            if(this.lastMessageId) path += "?since=" + this.lastMessageId;
            var request = this.client.request('GET', path, {
                host: this.host,
                authorization: "Basic " + this.base64AuthString
            });
            request.end();
            request.on('response',this.bind(this.messageCallback));
        }
        ,writable: false
        ,enumerable: false
        ,configurable: false
    },
    messageCallback: {
        value: function(response){
            response.setEncoding('utf8');

            var responseData = "";
            response.on('data', function(chunk){
                responseData += chunk;
            });

            scope = this;
            response.on('end', function(){
                scope.parseResponse(responseData);
            });
        }
        ,writable: false
        ,enumerable: false
        ,configurable: false
    }
    ,parseResponse: {
        value: function(responseData){
            var messages = JSON.parse(responseData);
            if(messages.length){
                for(var i = 0, len = messages.length; i < len; i++){
                    this.emit(messages[i].message.type, messages[i]);
                }
                this.lastMessageId = messages[len - 1].id;
            }
        }
        ,writable: false
        ,enumerable: false
        ,configurable: false
    }
});

module.exports = BackplanePoller;
